8Knot/pages/index/search_utils.py

```diff
@@ -0,0 +1,248 @@
+"""
+Search utilities for the 8Knot application.
+Provides improved search algorithms for the searchbar.
+"""
+from typing import List, Dict, Any
+import re
+from difflib import SequenceMatcher
+
+def levenshtein_distance(s1: str, s2: str) -> int:
+    """
+    Calculate the Levenshtein distance between two strings.
+    
+    The Levenshtein distance is the minimum number of single-character edits 
+    (insertions, deletions, or substitutions) required to change one string into another.
+    
+    Args:
+        s1: First string
+        s2: Second string
+        
+    Returns:
+        int: The Levenshtein distance between s1 and s2
+    """
+    # Convert both strings to lowercase for case-insensitive matching
+    s1, s2 = s1.lower(), s2.lower()
+    
+    # Create a matrix of size (len(s1)+1) x (len(s2)+1)
+    # Where matrix[i][j] will hold the Levenshtein distance between
+    # the first i characters of s1 and the first j characters of s2
+    rows, cols = len(s1) + 1, len(s2) + 1
+    dist = [[0 for _ in range(cols)] for _ in range(rows)]
+    
+    # Initialize the first row and column with the distance from empty string
+    for i in range(rows):
+        dist[i][0] = i
+    for j in range(cols):
+        dist[0][j] = j
+    
+    # Fill in the rest of the matrix
+    for i in range(1, rows):
+        for j in range(1, cols):
+            # If characters match, cost is 0, otherwise 1
+            cost = 0 if s1[i-1] == s2[j-1] else 1
+            
+            # Calculate minimum of three possible operations
+            dist[i][j] = min(
+                dist[i-1][j] + 1,      # deletion
+                dist[i][j-1] + 1,      # insertion
+                dist[i-1][j-1] + cost  # substitution
+            )
+    
+    # Return the distance
+    return dist[rows-1][cols-1]
+
+def levenshtein_score(s1: str, s2: str) -> float:
```

Comment:
https://pypi.org/project/python-Levenshtein/ - ratio from this package might be a more efficent option